### ISSUE TITLE
Reduce the size of the transaction data

### ISSUE THREAD

**Issue by robdefeo** [mailchain/mailchain#108]
**Is your feature request related to a problem? Please describe.**
Every byte costs on a transaction. The current transaction data has room for improvements. 
- MessageID default is 32 bytes
- Location is stored as encrypted string including, scheme, domain.
- The prefix is mailchain what is a shorter but still discoverable prefix.

**Describe the solution you'd like**
Replace or allow alternative transaction data format to optimise for byte usage.

**Describe alternatives you've considered**
Current object
```
message Data {
  int32 version = 1;
  bytes encryptedLocation = 2;
  bytes hash = 3;
}
```

**Additional context**

**Comment by robdefeo** [comment]
Transaction data is a difficult problem to get right as changes can introduce breaking changes, flexibility in the data structure helps with this although each byte costs when sending messages.

## Cost breakdown

- "Mailchain" prefix - 9 bytes - Replace with something shorter ideally distinguished single byte `0xCC` `0xFF` (get it merged into multihash)
- Multihash prefix - 1 byte
- Version - 2 bytes
- EncryptedHash - 8 bytes
- MessageID - 36 bytes
- DecryptedHash - 46 bytes (encryption increases size by 10 bytes)
- Location - variable

## Example data stores

Excluding the mailchain prefix and the single byte to describe the encoding (10 bytes total)

### Current

Total bytes is 172. Using the shortened mcx.mx domain. Encrypted location is efficient, and the version does not immediately offer any value.

```
message DataV0 {
  int32 version = 1;
  bytes encryptedLocation = 2;
  bytes hash = 3;
}
```

### Proposal 1 - Make schema a byte

Make schema a byte e.g. https, http, ipfs, other?

Total bytes is 107. Using shortened mcx.mx domain.

Question? What value does message ID offer?

```
message DataV2 {
  uint32 version = 1;
  bytes schema = 2;
  bytes host = 3;
  bytes messageId = 4;
  bytes encryptedHash = 5;
  bytes decryptedHash = 6;
}
```

### Proposal 2 - Remove messageID

Total bytes is 71. Using shortened mcx.mx domain.

```
message DataV3 {
    uint32 version = 1;
    bytes schema = 2;
    bytes host = 3;
    bytes decryptedHash = 4; 
    bytes encryptedHash = 5;
}
```

### Proposal 3 - Optional DNS

Total bytes is 64. Using shortened mcx.mx domain.

```
message DataV3DNS {
  uint32 version = 1;
  bytes schema = 2;
  bytes host = 3;
  bytes dns = 4;
  bytes decryptedHash = 5;
  bytes encryptedHash = 6;
}
```

### Proposal 4 - Remove version and combine schema and DNS (optional)

Where sending from a registered domain the size is 59 bytes, 64 bytes if using a domain like mcx.mx (6 chars)

```
message DataV4 {
  bytes schema = 1;
  bytes host = 2;
  bytes schemaDNS = 3;  
  bytes decryptedHash = 4;
  bytes encryptedHash = 5;
}
```

### Proposal 5 - Make schema and DNS int

Where sending from a registered domain the size is 58 bytes, 63 bytes if using a domain like mcx.mx (6 chars)
Int vs []byte for smaller numbers its more efficient

```
message DataV4A {
  uint32 schema = 1;
  bytes host = 2;
  uint64 schemaDNS = 3;  
  bytes decryptedHash = 4;
  bytes encryptedHash = 5;
}
```

**Comment by tboeckmann** [comment]
**Re: Question? What value does message ID offer?**

The current Mailchain [ethereum spec](https://github.com/mailchain/mailchain/tree/master/internal/chains/ethereum) functions as follows:

* Message ID: Is a unique message id composed of 64 chars (32 bytes) + @mailchain.
* The `message_filename` is stored as: `message-id`-`encrypted_message_hash`.
* The `message_location` is https://`storage_url`/`message_filename`.
* The `message_location` is encrypted and added to the transaction data.

The `message_location` is encrypted in txData for the recipient's eyes only.

The recipient can validate the message using the `encrypted_message_hash`.

So, message-id only serves to guarantee unique URLs, avoiding possible collisions with the addition of the `encrypted_message_hash`.

**Can the message ID be removed?**

Yes. If the txData contains a way to guarantee unique URLs within an address space.

**Comment by robdefeo** [comment]
The message ID is stored as a multihash which adds 2 bytes, 1 for representation of it as an ID, 1 for length. There is no value in this for the message ID stored as transaction as messageID randomness is needed and the first 2 bytes are always the same.

**Comment by robdefeo** [comment]
```
message Data {
  uint64 msgStoreCode = 1;  
  uint64 kind = 2;
  bytes host = 3;
  bytes resource = 4;
  bytes decryptedHash = 5;
  bytes encryptedHash = 6;
}
```

Data structure MUST contain the location of the message and SHOULD encrypt the location. 
Data structure MUST provide a means to verify its decrypted contents. 
Data structure SHOULD be efficient regarding usage of bytes.
A receipent MUST be able to decrypt the location of the message AND§ verify the decrypted data. 
An other party than the recipient MAY verify contents. 
`EncryptedHash` hash MAY be provided to veriy message data before decyption. Receipent MUST use `encryptedHash` if provided.

Data structure MUST contain either `msgStoreCode` or `kind` AND `host`

**Comment by robdefeo** [comment]
ref #217 ref #219 ref #218 ref #216 ref #215

**Comment by robdefeo** [comment]
closed by  #222 #223 #224
